tiny_dnn/parameter.h

```diff
@@ -0,0 +1,138 @@
+/*
+    Copyright (c) 2013, Taiga Nomi and the respective contributors
+    All rights reserved.
+
+    Use of this source code is governed by a BSD-style license that can be found
+    in the LICENSE file.
+*/
+#pragma once
+
+#include "tiny_dnn/core/framework/tensor.h"
+#include "tiny_dnn/util/util.h"
+
+namespace tiny_dnn {
+
+enum class parameter_type : int8_t { weight = 0x1, bias = 0x2 };
+
+class Parameter : public std::enable_shared_from_this<Parameter> {
+ public:
+  /**
+   * Initializes an empty parameter taking in the dimensions of weights and
+   * biases of a layer. Currently supported for maximum 4-dimensions, and
+   * stored as a flat ``Tensor``. Parameters are flat and represented in NCHW
+   * format.
+   *
+   * todo (karandesai) : generalize to n-dimensions
+   * todo (karandesai) : add an n-dimensional view for easy indexing
+   *
+   * @param out_channels number of feature maps in next layer
+   * @param in_channels  filter depth / input channels
+   * @param height       filter height
+   * @param width        filter width
+   * @param type         whether parameter is a weight or a bias
+   * @param trainable    whether parameter will be updated while training
+   */
+  // TODO: what is preferred order? probably, height and width first
+  Parameter(size_t out_channels,
+            size_t in_channels,
+            size_t height,
+            size_t width,
+            parameter_type type,
+            bool trainable = true)
+    : type_(type),
+      shape_(width, height, in_channels),
+      out_channels_(out_channels),
+      trainable_(trainable),
+      initialized_(false),
+      data_({shape_.size() * out_channels}),
+      grad_({1, shape_.size() * out_channels}) {}
+
+  // copy constructor
+  Parameter(const Parameter &other)
+    : type_(other.type()),
+      shape_(other.shape()),
+      out_channels_(other.size() / other.shape().size()),
+      trainable_(other.is_trainable()),
+      initialized_(other.initialized()),
+      data_(*(other.data())),
+      grad_(*(other.grad())) {}
+
+  shape3d shape() const { return shape_; }
+
+  size_t size() const { return data_.size(); }
+
+  parameter_type type() const { return type_; }
+
+  bool is_trainable() const { return trainable_; }
+
+  void set_trainable(bool trainable = true) { trainable_ = trainable; }
+
+  bool initialized() const { return initialized_; }
+
+  void set_initialized(bool initialized = true) { initialized_ = initialized; }
+
+  Tensor<float_t> *data() { return &data_; }
```

Comment:
@edgarriba I had that earlier, we shifted to shared_ptrs to avoid hassles of momery management. We can bring up this discussion separately and think about the possibilities. Luckily it is not a blocker of this PR ;)